app/uk/gov/hmrc/agentservicesaccount/auth/PasscodeVerification.scala

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright 2017 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.agentservicesaccount.auth
+
+import javax.inject.Inject
+
+import play.api.mvc.Results._
+import play.api.mvc.{Request, Result}
+import play.api.{Configuration, Environment, Mode}
+import uk.gov.hmrc.auth.otac.{Authorised, OtacAuthConnector}
+import uk.gov.hmrc.http.{HeaderCarrier, SessionKeys}
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class PasscodeVerificationException(msg: String) extends RuntimeException(msg)
+
+trait PasscodeVerification {
+  def apply[A](body: Boolean => Future[Result])(implicit request: Request[A], headerCarrier: HeaderCarrier, ec: ExecutionContext): Future[Result]
+}
+
+class FrontendPasscodeVerification @Inject()(configuration: Configuration,
+                                             environment: Environment,
+                                             otacAuthConnector: OtacAuthConnector)
+  extends PasscodeVerification {
+
+  val tokenParam = "p"
+  val passcodeEnabledKey = "passcodeAuthentication.enabled"
+  val passcodeRegimeKey = "passcodeAuthentication.regime"
+
+  lazy val passcodeEnabled: Boolean = configuration.getBoolean(passcodeEnabledKey).getOrElse(throwConfigNotFound(passcodeEnabledKey))
+  lazy val passcodeRegime: String = configuration.getString(passcodeRegimeKey).getOrElse(throwConfigNotFound(passcodeRegimeKey))
+  lazy val env: String = if (environment.mode.equals(Mode.Test)) "Test" else configuration.getString("run.mode").getOrElse("Dev")
+  lazy val verificationURL: String = configuration.getString(s"govuk-tax.$env.url.verification-frontend.redirect").getOrElse("/verification")
+  lazy val logoutUrl = s"$verificationURL/otac/logout/$passcodeRegime"
+
+  def loginUrl[A](queryParam: String) = s"$verificationURL/otac/login$queryParam"
+
+  def tokenQueryParam[A](implicit request: Request[A]): Option[String] =
+    request.getQueryString(tokenParam).map(token => s"?$tokenParam=$token")
+
+  def throwConfigNotFound(configKey: String) = throw new PasscodeVerificationException(s"The value for the key '$configKey' should be setup in the config file.")
+
+  def addRedirectUrl[A](implicit request: Request[A]): Result => Result = e =>
+    e.addingToSession(SessionKeys.redirect -> buildRedirectUrl(request))
+
+  def buildRedirectUrl[A](req: Request[A]): String =
+    if (env != "Prod") s"http${if (req.secure) "s" else ""}://${req.host}${req.path}" else req.path
```

Comment:
No tests assert on the result of this function. Do we need a '/' between host and path?

Reply:
thinking about it path will be prefixed with a slash, but we should really have a test checking the session value for both Dev and Prod scenarios

Reply:
good questions, my only excuse is that I copy-pasted it from official HMRC codebase of passcode-verification